cmd/otel-allocator/server/server.go

```diff
@@ -93,38 +85,36 @@ func (s *Server) Shutdown(ctx context.Context) error {
 	return s.server.Shutdown(ctx)
 }
 
-// ScrapeConfigsHandler returns the available scrape configuration discovered by the target allocator.
+// UpdateScrapeConfigResponse updates the scrape config response and potential error resulting from an update.
 // The target allocator first marshals these configurations such that the underlying prometheus marshaling is used.
 // After that, the YAML is converted in to a JSON format for consumers to use.
-func (s *Server) ScrapeConfigsHandler(c *gin.Context) {
-	configs := s.discoveryManager.GetScrapeConfigs()
-
-	hash, err := hashstructure.Hash(configs, nil)
+func (s *Server) UpdateScrapeConfigResponse(configs map[string]*promconfig.ScrapeConfig) {
+	var configBytes []byte
+	configBytes, err := yaml.Marshal(configs)
 	if err != nil {
-		s.logger.Error(err, "failed to hash the config")
-		s.errorHandler(c.Writer, err)
+		s.scrapeConfigErr = err
 		return
 	}
-	// if the hashes are different, we need to recompute the scrape config
-	if hash != s.compareHash {
-		var configBytes []byte
-		configBytes, err = yaml.Marshal(configs)
-		if err != nil {
-			s.errorHandler(c.Writer, err)
-			return
-		}
-		var jsonConfig []byte
-		jsonConfig, err = yaml2.YAMLToJSON(configBytes)
-		if err != nil {
-			s.errorHandler(c.Writer, err)
-			return
-		}
-		s.scrapeConfigResponse = jsonConfig
-		s.compareHash = hash
+	var jsonConfig []byte
+	jsonConfig, err = yaml2.YAMLToJSON(configBytes)
+	if err != nil {
+		s.scrapeConfigErr = err
+		return
+	}
+	s.scrapeConfigErr = nil
+	s.scrapeConfigResponse = jsonConfig
+}
+
+// ScrapeConfigsHandler returns the available scrape configuration discovered by the target allocator.
+func (s *Server) ScrapeConfigsHandler(c *gin.Context) {
+	response, err := s.scrapeConfigResponse, s.scrapeConfigErr
+	if err != nil {
+		s.errorHandler(c.Writer, s.scrapeConfigErr)
```

Comment:
Can async reading `s.scrapeConfigErr` here and writing it in ` func UpdateScrapeConfigResponse` lead to some issues?

Reply:
Good catch, I think actually I should put `err` here instead from `s.scrapeConfigErr`. That should ensure even if we change / remove the error cached on the HTTP server in the meantime, we won't put incorrect / nil error into the response.